### ISSUE TITLE
MetricsBucket is not thread safe (NullpointerException)

### ISSUE THREAD

**Issue by aleksabl** [Unleash/unleash-client-java#73]
We sometimes get this exception:

```
java.lang.NullPointerException: null
    at no.finn.unleash.metric.MetricsBucket.registerCount(MetricsBucket.java:19)
    at no.finn.unleash.metric.UnleashMetricServiceImpl.count(UnleashMetricServiceImpl.java:43)
    at no.finn.unleash.DefaultUnleash.count(DefaultUnleash.java:138)
    at no.finn.unleash.DefaultUnleash.isEnabled(DefaultUnleash.java:85)
```

The issue seems to be that MetricsBucket is not thread safe. Changing the code to use a `ConcurrentHashMap` (using `ConcurrentHashMap.computeIfAbsent` in `MetricsBucket.getOrCreate`) would likely fix the problem

**Comment by ivarconr** [comment]
Thanks for reporting. I will look in to it. 

This surprises me a bit sense the unleash SDK tries to only use a single dedicated thread for metrics handling.

How many instances of the unleash client do you instantiate in your application?

**Comment by aleksabl** [comment]
We only have a single instance. If you look at the stack trace you'll see pretty clearly that DefaultUnleash.isEnabled calls count (which ends up in MetricsBucket.registerCount in the same thread). I haven't looked much at the code, but maybe count should have been done through the event dispatcher (at ToggleEvaluated)?

**Comment by ivarconr** [comment]
Thanks, you are absolutely right, `ConcurrentHashMap` is the correct solution in this scenario. 

Multiple threads knocking on `MetricsBucket.registerCount()` could from time to time trigger concurrency issues (even though I cant find any traces of this happening across our java apps. 

released a fix as `3.2.4`.

**Comment by aleksabl** [comment]
Hi, thanks for looking into it. I had a look at the code and I still found several issues regarding thread safety.

I'll try to fix the issues I found and send a pull request today.
